src/main/scala/rocket/RVC.scala

```diff
@@ -95,16 +119,50 @@ class RVCDecoder(x: UInt, xLen: Int, useAddiForMv: Boolean = false) {
     def beqz = inst(Cat(bImm(12), bImm(10,5), x0, rs1p, 0.U(3.W), bImm(4,1), bImm(11), 0x63.U(7.W)), rs1p, rs1p, x0)
     def bnez = inst(Cat(bImm(12), bImm(10,5), x0, rs1p, 1.U(3.W), bImm(4,1), bImm(11), 0x63.U(7.W)), x0, rs1p, x0)
     def arith = {
-      def srli = Cat(shamt, rs1p, 5.U(3.W), rs1p, 0x13.U(7.W))
-      def srai = srli | (1 << 30).U
-      def andi = Cat(addiImm, rs1p, 7.U(3.W), rs1p, 0x13.U(7.W))
+      def srli = inst(Cat(shamt, rs1p, 5.U(3.W), rs1p, 0x13.U(7.W)), rs1p, rs1p, rs2p)
+      def srai = inst(Cat(0x10.U, shamt, rs1p, 5.U(3.W), rs1p, 0x13.U(7.W)), rs1p, rs1p, rs2p)
+      def andi = inst(Cat(addiImm, rs1p, 7.U(3.W), rs1p, 0x13.U(7.W)), rs1p, rs1p, rs2p)
       def rtype = {
-        val funct = Seq(0.U, 4.U, 6.U, 7.U, 0.U, 0.U, 2.U, 3.U)(Cat(x(12), x(6,5)))
+        val funct = Seq(0.U, 4.U, 6.U, 7.U, 0.U, 0.U, 0.U, 3.U)(Cat(x(12), x(6,5)))
         val sub = Mux(x(6,5) === 0.U, (1 << 30).U, 0.U)
-        val opc = Mux(x(12), 0x3B.U(7.W), 0x33.U(7.W))
-        Cat(rs2p, rs1p, funct, rs1p, opc) | sub
+        val opc = Mux(x(12), Mux(x(6), 0x33.U(7.W), 0x3B.U(7.W)), 0x33.U(7.W))
+        def mul = {
+          if(usingMulDiv && usingCompressedSuiteB) Mux(Cat(x(12), x(6,5)) === 6.U, (1 << 25).U, 0.U)
+          else 0.U
+        }
+        def zca = inst(Cat(rs2p, rs1p, funct, rs1p, opc) | sub | mul, rs1p, rs1p, rs2p)
+        def zcb_q1 = {
+          def unimp = inst(Cat(lwImm >> 5, rs2p, rs1p, 2.U(3.W), lwImm(4,0), 0x3F.U(7.W)), rs2p, rs1p, rs2p)
+          if(usingCompressedSuiteB){
+            def zextb = inst(Cat(0xff.U, rs1p, 7.U(3.W), rs1p, 0x13.U(7.W)), rs1p, rs1p, rs2p)
+            def not   = inst(Cat(0xFFF.U, rs1p, 4.U(3.W), rs1p, 0x13.U(7.W)), rs1p, rs1p, rs2p)
+            def zexth64 = inst(Cat(0x80.U, rs1p, 4.U(3.W), rs1p, 0x3B.U(7.W)), rs1p, rs1p, rs2p)
+            def sextb = {
+              if(usingBitManip) inst(Cat(0x604.U, rs1p, 1.U(3.W), rs1p, 0x13.U(7.W)), rs1p, rs1p, rs2p)
+              else unimp
+            }
+            def sexth = {
+              if(usingBitManip) inst(Cat(0x605.U, rs1p, 1.U(3.W), rs1p, 0x13.U(7.W)), rs1p, rs1p, rs2p)
+              else unimp
+            }
+            def zextw = {
+              if(usingBitManip) inst(Cat(4.U, x0, rs1p, 0.U(3.W), rs1p, 0x3B.U(7.W)), rs1p, rs1p, x0)
+              else unimp
+            }
+            def zexth = {
+              if(usingBitManip) {
+                if (xLen == 32) inst(Cat(0x80.U, rs1p, 4.U(3.W), rs1p, 0x33.U(7.W)), rs1p, rs1p, rs2p)
+                else zexth64
+              }
+              else unimp
+            }
+            Seq(zextb, sextb, zexth, sexth, zextw, not)(x(4,2))
+          }
+          else unimp
+        }
+        Mux(Cat(x(12), x(6,5)) === 7.U, zcb_q1, zca)
       }
-      inst(Seq(srli, srai, andi, rtype)(x(11,10)), rs1p, rs1p, rs2p)
+      Seq(srli, srai, andi, rtype)(x(11,10))
```

Comment:
> Is there any special reason for it? The x0 in zextw?

Yes, this is the reason.

> the user will have a larger area even the user does not enable usingCompressedSuiteB

I'm a little skeptical about it, and I have following two reasons to back my argument:
1. As per my (vague) understanding of hardware (Verilog) generation using CHISEL, computation is done in software by keeping into account the parameters, and then the Verilog is generated including (only) that hardware part that was allowed by the parameter.
2. The above can be verified by observing the generated Verilog (by looking at the reference at the end of each line of generated Verilog that points to the corresponding scala file and its line number). For example, when I build Rocket with `usingCompressedSuiteB = false`, the generated Verilog simply excludes the corresponding definition of 16->32 instructions. But when I build using `usingCompressedSuiteB = true`, generated Verilog does include the references of the definition of those instructions which are defined under the if clause of this parameter.

So, this should not increase the area if `usingCompressedSuiteB` is disabled, unless I'm missing something.

Reply:
Further details should be checked. There are two steps of selection, static selection like if/elses in scala and dynamic selection like `x(11,10)` in the circuit.

The DefaultConfig in dev branch

```verilog
// the final mux
  assign io_out_bits = _GEN_19[_io_out_T_2];
  assign io_out_rd = _GEN_20[_io_out_T_2];
  assign io_out_rs1 = _GEN_21[_io_out_T_2];
  assign io_out_rs2 = _GEN_22[_io_out_T_2];
  assign io_out_rs3 = _GEN_23[_io_out_T_2];
  
// lets check rs1
  wire [31:0][4:0]  _GEN_21 =
  ...
    {io_out_s_12_rs1}, // this coresponds to arith in scala
  ...
  
  assign io_out_s_12_rs1 = {2'h1, io_in[9:7]};
```

DefaultConfig in this branch

```verilog
// the final mux
  assign io_out_bits = _GEN_23[_io_out_T_2];
  assign io_out_rd = _GEN_24[_io_out_T_2];
  assign io_out_rs1 = _GEN_25[_io_out_T_2];
  assign io_out_rs2 = _GEN_26[_io_out_T_2];
  assign io_out_rs3 = _GEN_27[_io_out_T_2];
  
// lets check rs1
  wire [31:0][4:0]  _GEN_25 =
  ...
    {io_out_s_12_rd}, // this coresponds to arith in scala
  ...
  
  wire [4:0]        io_out_s_12_rs1 = _GEN_8[io_in[11:10]]; // extra mux
  
  wire [3:0][4:0]   _GEN_8 =
    {{(&_io_out_s_T_235) ? io_out_s_res_4_rs1 : io_out_s_res_5_rs1},
     {io_out_s_res_3_rs1},
     {io_out_s_res_2_rs1},
     {io_out_s_res_1_rs1}};
     
  assign io_out_s_res_4_rs1 = {2'h1, io_in[9:7]};
  assign io_out_s_res_5_rs1 = {2'h1, io_in[9:7]};
  assign io_out_s_res_3_rs1 = {2'h1, io_in[9:7]};
  assign io_out_s_res_2_rs1 = {2'h1, io_in[9:7]};
  assign io_out_s_res_1_rs1 = {2'h1, io_in[9:7]};
```

Chisel is faithful of generating what you write. It might be argued that the synthesizer will optimize this out, but I think avoiding this from source is a better option.

Reply:
Thanks for the detailed explanation, I understand now. Pushed fixes in the latest commit.